crates/uv/src/commands/python/pin.rs

```diff
@@ -82,3 +122,102 @@ pub(crate) async fn pin(
 
     Ok(ExitStatus::Success)
 }
+
+/// Check if pinned request is compatible with the workspace/project's `Requires-Python`.
+fn check_request_requires_python_compatibility(
+    pin: &PythonRequest,
+    virtual_project: &VirtualProject,
+    python_preference: PythonPreference,
+    cache: &Cache,
+) {
+    let requested_version = match pin {
+        PythonRequest::Version(ref version) => {
```

Comment:
We also need to handle `PythonRequest::InterpreterVersion` and `PythonRequest::Key` when there's a version present.